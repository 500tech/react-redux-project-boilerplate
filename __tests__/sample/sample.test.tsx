import React from 'react';
import { shallow, mount } from 'enzyme';

import { Props, Sample, StyledPost } from 'sample/sample';
import { act } from 'react-dom/test-utils';

// mock the injection of the reducer into the application as we don't run
// tests with any reference to the store
// this is only done for sample and should not be used in a regular application
jest.mock('sample/sample-replace-reducer', () => {});

const setup = ({
  posts = [],
  isLoading = false,
  fetchPosts = jest.fn(),
  setLocale = jest.fn()
}: Partial<Props> = {}) => ({
  posts,
  isLoading,
  fetchPosts,
  setLocale
});

describe('<Sample />', () => {
  test('should render', () => {
    const props = setup();
    const component = shallow(<Sample {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('should render loading when component is in loading state', () => {
    const props = setup({ isLoading: true });
    const component = shallow(<Sample {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('should render 3 posts', () => {
    const posts = [
      { id: 1, title: 'first post', body: 'first post content', userId: 1 },
      { id: 2, title: 'second post', body: 'second post content', userId: 2 },
      { id: 3, title: 'third post', body: 'third post content', userId: 3 }
    ];
    const props = setup({ posts });
    const component = shallow(<Sample {...props} />);

    const postsCount = component.find(StyledPost).length;

    expect(postsCount).toEqual(3);
    expect(component).toMatchSnapshot();
  });

  test('should fetch posts when mounting', () => {
    const props = setup({ fetchPosts: jest.fn() });
    mount(<Sample {...props} />);

    expect(props.fetchPosts).toHaveBeenCalled();
  });

  test('should fetch posts when clicking refresh button', () => {
    const props = setup({ fetchPosts: jest.fn() });
    const component = mount(<Sample {...props} />);

    // We check that fetch posts have been called once already as it is run in componentDidMount
    expect(props.fetchPosts).toHaveBeenCalledTimes(1);

    act(() => {
      component.find('[alt="refresh"]').simulate('click');
      expect(props.fetchPosts).toHaveBeenCalledTimes(2);
    });
  });
});
