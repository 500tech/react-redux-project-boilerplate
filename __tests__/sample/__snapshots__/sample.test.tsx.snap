// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sample /> should render 1`] = `
<StyledContainer>
  <h1>
    Title
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div />
</StyledContainer>
`;

exports[`<Sample /> should render 3 posts 1`] = `
<StyledContainer>
  <h1>
    Title
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div>
    <StyledPost
      key="1"
    >
      <h4>
        first post
      </h4>
      <p>
        first post content
      </p>
    </StyledPost>
    <StyledPost
      key="2"
    >
      <h4>
        second post
      </h4>
      <p>
        second post content
      </p>
    </StyledPost>
    <StyledPost
      key="3"
    >
      <h4>
        third post
      </h4>
      <p>
        third post content
      </p>
    </StyledPost>
  </div>
</StyledContainer>
`;

exports[`<Sample /> should render loading when component is in loading state 1`] = `
<StyledContainer>
  <h1>
    Title
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div>
    loading...
  </div>
</StyledContainer>
`;
